README.md
# ZFontIcon - Icon fonts support for Qt applications
ZFontIcon is released under the MIT License


## Description
ZFontIcon provides a simple way to use any glyph/icon font as QIcon in your Qt5/Qt6 applications.

Some helpers are included to use the following icon fonts with no effort:
- Font Awesome 6.4.2  (Free, Brands and Pro + Sharp)
- Font Awesome 5.15.4 (Free, Brands and Pro)
- Font Awesome 4.7.0
- Bootstrap Icons 1.10.5
- Google Material Design Icons 4 (github master 511eea5 - 2023-01-07)
- Material Design Icons 7.2.96
- Phosphor 2.0
- Phosphor 1.4

You can use any other icon fonts using Unicode character codes. For convenience, I suggest you create your own font helper: simply create an enumeration listing all the Unicode characters of the font icons. Have a look at the code, it's very easy!


## Qt Compatibility
ZFontIcon is tested under Windows, and Linux with:
- Qt 6.5  LTS
- Qt 6.2  LTS
- Qt 5.15 LTS


## Installation
The easiest way to include ZFontIcon in your project is to copy the ZFontIcon folder to your project tree and add the following include() to your Qt project (.pro) file:

    include("ZFontIcon/ZFontIcon.pri")

Now, you are ready!


## Usage
### Basic usage

First of all, you need to load the icon fonts you want to use (typically in the main.cpp):

```c++
#include "ZFontIcon/ZFontIcon.h"
#include "ZFontIcon/ZFont_fa6.h" // FA6 helpers

// FA6 Free
ZFontIcon::addFont(":/fa6/" + Fa6::FA6_TTF_FILE_FREE_SOLID);
ZFontIcon::addFont(":/fa6/" + Fa6::FA6_TTF_FILE_FREE_REGULAR);

// FA6 Pro (if you have a license)
ZFontIcon::addFont(":/fa6/" + Fa6pro::FA6_TTF_FILE_PRO_SOLID);
ZFontIcon::addFont(":/fa6/" + Fa6pro::FA6_TTF_FILE_PRO_REGULAR);
ZFontIcon::addFont(":/fa6/" + Fa6pro::FA6_TTF_FILE_PRO_LIGHT);
ZFontIcon::addFont(":/fa6/" + Fa6pro::FA6_TTF_FILE_PRO_THIN);

// FA6 Pro Sharp (if you have a license)
ZFontIcon::addFont(":/fa6/" + Fa6sharp::FA6_TTF_FILE_SHARP_SOLID);
ZFontIcon::addFont(":/fa6/" + Fa6sharp::FA6_TTF_FILE_SHARP_REGULAR);
ZFontIcon::addFont(":/fa6/" + Fa6sharp::FA6_TTF_FILE_SHARP_LIGHT);

// FA6 Brands
ZFontIcon::addFont(":/fa6/" + Fa6brands::FA6_TTF_FILE_BRANDS);
```

Then you can get your favorite icons:

```c++
#include "ZFontIcon/ZFontIcon.h"
#include "ZFontIcon/ZFont_fa6.h"

// FA6 Free (FA6 Free include about 150 regular icon)
ZFontIcon::icon(Fa6::FAMILY,       Fa6::SOLID,      Fa6::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6::FAMILY,       Fa6::REGULAR,    Fa6::fa_flag, QColor(195,  65,  65));

// FA6 Pro (only if you have a license)
ZFontIcon::icon(Fa6pro::FAMILY,    Fa6pro::SOLID,   Fa6pro::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6pro::FAMILY,    Fa6pro::REGULAR, Fa6pro::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6pro::FAMILY,    Fa6pro::LIGHT,   Fa6pro::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6pro::FAMILY,    Fa6pro::THIN,    Fa6pro::fa_flag, QColor(195,  65,  65));

// FA6 Sharp (only if you have a license)
//   Use the glyphs defined in fa5pro namespace because they are identical to Fa6sharp glyphs
ZFontIcon::icon(Fa6sharp::FAMILY,  Fa6sharp::SOLID,   Fa6pro::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6sharp::FAMILY,  Fa6sharp::REGULAR, Fa6pro::fa_flag, QColor(195,  65,  65));
ZFontIcon::icon(Fa6sharp::FAMILY,  Fa6sharp::LIGHT,   Fa6pro::fa_flag, QColor(195,  65,  65));

// FA6 Brands (note that FA6 Brands has no font style)
ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_github, QColor(128, 0, 0));
```


### Intermediate usage
Need to set default icon color for the whole project? No problem, use the global definition!

```c++
ZFontIconOption::gColor=         QColor(195,  65,  65); // Default color
ZFontIconOption::gColorDisabled= QColor(122, 196,  62); // Color when icon is disabled
ZFontIconOption::gColorSelected= QColor(122, 196,  62); // Color when icon is selected

// Optional settings: if not set, color value will be used
ZFontIconOption::gColorOn=       QColor(102, 163,  52); // Color when icon state is On
ZFontIconOption::gColorActive=   QColor(234,  78,  78); // Color when icon is hovered
ZFontIconOption::gColorActiveOn= QColor(122, 196,  62); // Color when icon is hovered and state is On
```


### Advanced usage
More customization needed? You can set all aspects of your icon:

- Rotate, flip from top to bottom and flip from left to right icon
- On icon state change [On/Off] you can use different:
  - font styles (eg. switch from Solid to Regular font style);
  - glyphs (eg. change glyph from lock to unlock);
  - colors (override global colors);
  - scale factors (eg. make On state glyph bigger than Off state glyph);
- On mouse hover, you can:
  - set a hover color for Off state;
  - set a hover color for On state,
- For icon mode Selected and Disabled, you can change the default color.

```c++
ZFontIconOption fIcon;

// Minimum required arguments
fIcon.fontFamily=    Fa6::FAMILY;           // Set the font icon family
fIcon.fontStyle=     Fa6::SOLID;            // Set the family style (Required if font families have multiple registered styles)
fIcon.glyph=         Fa6::fa_grin;          // Set the default glyph
fIcon.color=         QColor(195,  65,  65); // Set the default color

// Optional arguments
fIcon.fontStyleOn=   Fa6::REGULAR;          // Font style when icon state is On
fIcon.glyphOn=       Fa6::fa_grin_stars;    // Glyph when icon state is On

fIcon.colorOn=       QColor(102, 163,  52); // Color when icon state is On
fIcon.colorActive=   QColor(234,  78,  78); // Color when icon is hovered
fIcon.colorActiveOn= QColor(122, 196,  62); // Color when icon is hovered and state is On
fIcon.colorDisabled= QColor(122, 196,  62); // Color when icon is disabled
fIcon.colorSelected= QColor(122, 196,  62); // Color when icon is selected

fIcon.scaleFactor=   0.80;                  // Default icon scale factor
fIcon.scaleFactorOn= 1.00;                  // Scale factor when icon state is On

fIcon.rotateAngle=   0.00;                  // Icon clockwise rotation in degree
fIcon.flipLeftRight= false;                 // Icon flip from left to right
fIcon.flipTopBottom= false;                 // Icon flip from top to bottom

ZFontIcon::icon(fIcon);
```


## Tips
With meta-object support for font helper namespaces, you can iterate over the namespace:

```c++
QMetaEnum metaEnum= QMetaEnum::fromType<Fa6::Glyph>();
for(int i= 0; i < metaEnum.keyCount(); ++i) {
    qDebug() << metaEnum.key(i);   // Name of the given index
    qDebug() << metaEnum.value(i); // Value with the given index
}
```


## Example
The following repository contains a commented example based on Font Awesome.
Open ZFontIconSample.pro in QtCreator and run it or read sources for more information.
Check the comments at the beginning of main.cpp to enable FA6 Pro or FA5 Pro.


## Support
I do not provide any specific support, but if you have any question, please feel free to reach me out via email at philvl.dev(at)gmail.com


## Credits

CHANGELOG.md
### Release v2.2.3 - 2023-08-21
- Helpers
  - Update helper for Font Awesome 6.4.2

### Release v2.2.2 - 2023-06-11
- Helpers
  - Add helper for Phosphor Icons 2.0
  - Update helper for Bootstrap Icons 1.10.5
  - Update helper for Font Awesome 6.4.0 (Add Sharp Light)
  - Update helper for Material Design Icons 7.2.96
- ZFontIcon Sample Application
  - Phosphor v1.4 is not compatible with Phosphor v2.0, consequently in the demo application the integration
    of Phosphor has been updated to v2.0. (Phopsphor 1.4 fonts and helpers files has been keeped)
  - Add Website and GitHub link for each fonts

### Release v2.2.1 - 2023-02-17
- Helpers
  - Update helper for Bootstrap Icons 1.10.3
  - Update helper for Font Awesome 6.3.0 (Add Sharp Regular)
  - Update helper for Google Material Icons 4 (github master 511eea5 - 2023-01-07)
  - Update helper for Material Design Icons 7.1.96

### Release v2.2.0 - 2022-09-20
- ZFontIcon
  - Added global colors to easily set them for the whole project (Ti-R contrib. | https://github.com/Ti-R)
- Helpers
  - Update helper for Bootstrap Icons 1.9.1
  - Update helper for Font Awesome 6.2.0 (Add Sharp Solid)
  - Update helper for Google Material Icons 4 (v2022-08-01, github master 1c57576)
  - Update helper for Material Design Icons 6.9.96

### Release v2.1.1 - 2022-04-09
- Update helper for Font Awesome 6.1.1
- Update helper for Google Material Icons 4 (v2022-03-31, github master e55dd24)

### Release v2.1.0 - 2022-02-11
- ZFontIcon
  - Add new icon options: rotate, flipTopBottom and flipLeftRight
  - Minor code improvements
- Helpers
  - Add helper for Bootstrap Icons 1.8.1
  - Add helper for Material Design Icons 6.5.95
  - Add helper for Phosphor Icons 1.4
  - Update helper for Font Awesome 6.0.0
  - Update helper for Font Awesome 5.15.4
  - Update helper for Google Material Icons 4 (github master c1eb628 - 2022-01-07)
- Sample:
  - Add tabs for Bootstrap Icons
  - Add tabs for Material Design Icons
  - Add tabs for Phosphor Icons

### Release v2.0.10 - 2021-06-04
- Add UCS-4/UTF-32 unicode glyph support (4 bytes unicode character)
- Update helper for Font Awesome 6.0.0-beta1 (requires a FA pro licence to get alpha fonts)
- Update helper for Google Material Design Icons 4 (github master ab12f16 - 2021-05-11)

### Release v2.0.9 - 2021-04-01
- Update helper for Font Awesome 6.0.0-alpha3 (requires a FA pro licence to get alpha fonts)
- Update helper for Font Awesome 5.15.3

### Release v2.0.8 - 2021-01-08
- Update helper for Font Awesome 5.15.2

### Release v2.0.7 - 2020-12-25
- Update helper for Font Awesome 6.0.0-alpha2 (requires a FA pro licence to get alpha fonts)

### Release v2.0.6 - 2020-12-11
- BugFix: Qt 6.0.0 compatibility
  - Use QFontDatabase static functions
  - Resolve ambiguous glyph type: use ushort instead of QChar

### Release v2.0.5 - 2020-11-15
- Sample:
  - CheatSheet: add icon search engine
  - CheatSheet: show unicode value in tooltip
  - BugFix: executable recognizes as shared library under Ubuntu Linux likes

### Release v2.0.4 - 2020-11-09
- ZFontIcon
  - Add function isRegistered();
- Fonts
  - Bugfix: Google material icons not run on Windows
- Helpers:
  - Add helper for Font Awesome 6.0.0-alpha1 (requires a FA pro licence to get alpha fonts)
  - Add meta-object support for fonts helpers namespaces (you can now iterate over namespaces)
- Sample:
  - Add prewiew tabs for FA6
  - Add a font cheatsheet tool
- Add CHANGELOG.md file

### Release v2.0.3 - 2020-11-08

sample/Dialog.cpp
#include "Dialog.h"
#include "ui_Dialog.h"

// Font icons
#include "ZFontIcon/ZFontIcon.h"
#include "ZFontIcon/ZFont_fa6.h"
#include "ZFontIcon/ZFont_fa5.h"
#include "ZFontIcon/ZFont_fa4.h"
#include "ZFontIcon/ZFont_gmi4.h"

Dialog::Dialog(QWidget *parent) : QDialog(parent), ui(new Ui::Dialog) {
    ui->setupUi(this);
    ui->tabWidget->setCurrentIndex(0);

    // Show registered fonts (family and styles) in a treeWidget
    QMapIterator<QString, QStringList> iterator(ZFontIcon::registeredFonts());
    while (iterator.hasNext()) {
        iterator.next();
        QTreeWidgetItem *familyItem= new QTreeWidgetItem(ui->treeWidget);
        familyItem->setText(0, iterator.key());
        familyItem->setExpanded(true);
        for(const QString &style : iterator.value()) {
            QTreeWidgetItem *styleItem= new QTreeWidgetItem(familyItem);
            styleItem->setText(0, style);
        }
    }

    //--------------------------------
    // Font Awesome 6 FREE (Preview) -
    //--------------------------------
    if(ZFontIcon::registeredFonts().contains(Fa6::FAMILY) && ZFontIcon::registeredFonts().contains(Fa6brands::FAMILY)) {
        // The family of the first loaded font is used by default. In this example, the
        // "Font Awesome 6 Solid" font family is set by default. Solid, Regular, Light
        // and Thin styles share the same family name so specify the family is useless.
        // As different styles from the same font family have been loaded (Solid and Regular),
        // we need to specify which style to use explicitly
        ui->fa6_btn101->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::SOLID,   Fa6::fa_flag,    QColor(195,  65,  65)));
        ui->fa6_btn102->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::SOLID,   Fa6::fa_star,    QColor(214, 197,  64)));
        ui->fa6_btn103->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::SOLID,   Fa6::fa_user,    QColor( 69, 100, 214)));
        ui->fa6_btn104->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::SOLID,   Fa6::fa_map,     QColor(102, 163,  52)));
        ui->fa6_btn105->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::SOLID,   Fa6::fa_comment, QColor(  0,   0,   0)));

        ui->fa6_btn111->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::REGULAR, Fa6::fa_flag,    QColor(195,  65,  65)));
        ui->fa6_btn112->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::REGULAR, Fa6::fa_star,    QColor(214, 197,  64)));
        ui->fa6_btn113->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::REGULAR, Fa6::fa_user,    QColor( 69, 100, 214)));
        ui->fa6_btn114->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::REGULAR, Fa6::fa_map,     QColor(102, 163,  52)));
        ui->fa6_btn115->setIcon(ZFontIcon::icon(Fa6::FAMILY, Fa6::REGULAR, Fa6::fa_comment, QColor(  0,   0,   0)));

        // Advanced use exemple with Font Awesome 6 FREE
        ZFontIconOption fIconFa6;
        fIconFa6.fontFamily=    Fa6::FAMILY;
        fIconFa6.fontStyle=     Fa6::SOLID;
        fIconFa6.fontStyleOn=   Fa6::REGULAR;
        fIconFa6.glyph=         Fa6::fa_face_grin;
        fIconFa6.glyphOn=       Fa6::fa_face_grin_stars;
        fIconFa6.color=         QColor(195,  65,  65);
        fIconFa6.colorOn=       QColor(102, 163,  52);
        fIconFa6.colorActive=   QColor(234,  78,  78);
        fIconFa6.colorActiveOn= QColor(122, 196,  62);
        fIconFa6.scaleFactor=   0.85;
        fIconFa6.scaleFactorOn= 1.00;
        ui->fa6_btn121->setIcon(ZFontIcon::icon(fIconFa6));
        ui->fa6_btn122->setIcon(ZFontIcon::icon(fIconFa6));
        ui->fa6_btn123->setIcon(ZFontIcon::icon(fIconFa6));


        //----------------------------------
        // Font Awesome 6 BRANDS (Preview) -
        //----------------------------------
        // FA6 Brands is not the first added font, so we must specify the family explicitly
        // FA6 Brands font has only one style so we don't need to specify it
        ui->fa6_btn131->setIcon(ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_github    ));
        ui->fa6_btn132->setIcon(ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_gitlab    ));
        ui->fa6_btn133->setIcon(ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_facebook  ));
        ui->fa6_btn134->setIcon(ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_twitter   ));
        ui->fa6_btn135->setIcon(ZFontIcon::icon(Fa6brands::FAMILY, Fa6brands::fa_instagram ));
    }
    else {
        // Remove FA6 Free tab
        QWidget *fa6_tab= ui->tabWidget->findChild<QWidget*>("fa6_tab");
        int tabIndex= ui->tabWidget->indexOf(fa6_tab);
        ui->tabWidget->removeTab(tabIndex);
    }


    //-------------------------------
    // Font Awesome 6 PRO (Preview) -
    //-------------------------------
    if(ZFontIcon::registeredFonts().contains(Fa6pro::FAMILY)) {
        ui->fa6pro_btn201->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::SOLID,   Fa6pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa6pro_btn202->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::SOLID,   Fa6pro::fa_star,    QColor(214, 197,  64)));
        ui->fa6pro_btn203->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::SOLID,   Fa6pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa6pro_btn204->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::SOLID,   Fa6pro::fa_map,     QColor(102, 163,  52)));
        ui->fa6pro_btn205->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::SOLID,   Fa6pro::fa_comment, QColor(  0,   0,   0)));

        ui->fa6pro_btn211->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::REGULAR, Fa6pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa6pro_btn212->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::REGULAR, Fa6pro::fa_star,    QColor(214, 197,  64)));
        ui->fa6pro_btn213->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::REGULAR, Fa6pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa6pro_btn214->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::REGULAR, Fa6pro::fa_map,     QColor(102, 163,  52)));
        ui->fa6pro_btn215->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::REGULAR, Fa6pro::fa_comment, QColor(  0,   0,   0)));

        ui->fa6pro_btn221->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::LIGHT,   Fa6pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa6pro_btn222->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::LIGHT,   Fa6pro::fa_star,    QColor(214, 197,  64)));
        ui->fa6pro_btn223->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::LIGHT,   Fa6pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa6pro_btn224->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::LIGHT,   Fa6pro::fa_map,     QColor(102, 163,  52)));
        ui->fa6pro_btn225->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::LIGHT,   Fa6pro::fa_comment, QColor(  0,   0,   0)));

        ui->fa6pro_btn231->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::THIN,    Fa6pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa6pro_btn232->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::THIN,    Fa6pro::fa_star,    QColor(214, 197,  64)));
        ui->fa6pro_btn233->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::THIN,    Fa6pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa6pro_btn234->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::THIN,    Fa6pro::fa_map,     QColor(102, 163,  52)));
        ui->fa6pro_btn235->setIcon(ZFontIcon::icon(Fa6pro::FAMILY, Fa6pro::THIN,    Fa6pro::fa_comment, QColor(  0,   0,   0)));
    }
    else {
        // Remove FA6 Pro tab
        QWidget *fa6pro_tab= ui->tabWidget->findChild<QWidget*>("fa6pro_tab");
        int tabIndex= ui->tabWidget->indexOf(fa6pro_tab);
        ui->tabWidget->removeTab(tabIndex);
    }


    //----------------------
    // Font Awesome 5 FREE -
    //----------------------
    // The family of the first loaded font is used by default. In this example, the
    // "Font Awesome 5 Solid" font family is set by default. Solid, Regular, and Light
    // styles share the same family name so specify the family is useless.
    // As different styles from the same font family have been loaded (Solid and Regular),
    // we need to specify which style to use explicitly
    ui->fa5_btn101->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::SOLID,   Fa5::fa_flag,    QColor(195,  65,  65)));
    ui->fa5_btn102->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::SOLID,   Fa5::fa_star,    QColor(214, 197,  64)));
    ui->fa5_btn103->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::SOLID,   Fa5::fa_user,    QColor( 69, 100, 214)));
    ui->fa5_btn104->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::SOLID,   Fa5::fa_map,     QColor(102, 163,  52)));
    ui->fa5_btn105->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::SOLID,   Fa5::fa_comment, QColor(  0,   0,   0)));

    ui->fa5_btn111->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::REGULAR, Fa5::fa_flag,    QColor(195,  65,  65)));
    ui->fa5_btn112->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::REGULAR, Fa5::fa_star,    QColor(214, 197,  64)));
    ui->fa5_btn113->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::REGULAR, Fa5::fa_user,    QColor( 69, 100, 214)));
    ui->fa5_btn114->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::REGULAR, Fa5::fa_map,     QColor(102, 163,  52)));
    ui->fa5_btn115->setIcon(ZFontIcon::icon(Fa5::FAMILY, Fa5::REGULAR, Fa5::fa_comment, QColor(  0,   0,   0)));

    // Advanced use exemple with Font Awesome 5 FREE
    ZFontIconOption fIconFa5;
    fIconFa5.fontFamily=    Fa5::FAMILY;
    fIconFa5.fontStyle=     Fa5::SOLID;
    fIconFa5.fontStyleOn=   Fa5::REGULAR;
    fIconFa5.glyph=         Fa5::fa_grin;
    fIconFa5.glyphOn=       Fa5::fa_grin_stars;
    fIconFa5.color=         QColor(195,  65,  65);
    fIconFa5.colorOn=       QColor(102, 163,  52);
    fIconFa5.colorActive=   QColor(234,  78,  78);
    fIconFa5.colorActiveOn= QColor(122, 196,  62);
    fIconFa5.scaleFactor=   0.85;
    fIconFa5.scaleFactorOn= 1.00;
    ui->fa5_btn121->setIcon(ZFontIcon::icon(fIconFa5));
    ui->fa5_btn122->setIcon(ZFontIcon::icon(fIconFa5));
    ui->fa5_btn123->setIcon(ZFontIcon::icon(fIconFa5));


    //------------------------
    // Font Awesome 5 BRANDS -
    //------------------------
    // FA5 Brands is not the first added font, so we must specify the family explicitly
    // FA5 Brands font has only one style so we don't need to specify it
    ui->fa5_btn131->setIcon(ZFontIcon::icon(Fa5brands::FAMILY, Fa5brands::fa_github    ));
    ui->fa5_btn132->setIcon(ZFontIcon::icon(Fa5brands::FAMILY, Fa5brands::fa_gitlab    ));
    ui->fa5_btn133->setIcon(ZFontIcon::icon(Fa5brands::FAMILY, Fa5brands::fa_facebook  ));
    ui->fa5_btn134->setIcon(ZFontIcon::icon(Fa5brands::FAMILY, Fa5brands::fa_twitter   ));
    ui->fa5_btn135->setIcon(ZFontIcon::icon(Fa5brands::FAMILY, Fa5brands::fa_instagram ));


    //---------------------
    // Font Awesome 5 PRO -
    //---------------------
    if(ZFontIcon::registeredFonts().contains(Fa5pro::FAMILY)) {
        ui->fa5pro_btn201->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::SOLID,   Fa5pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa5pro_btn202->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::SOLID,   Fa5pro::fa_star,    QColor(214, 197,  64)));
        ui->fa5pro_btn203->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::SOLID,   Fa5pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa5pro_btn204->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::SOLID,   Fa5pro::fa_map,     QColor(102, 163,  52)));
        ui->fa5pro_btn205->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::SOLID,   Fa5pro::fa_comment, QColor(  0,   0,   0)));

        ui->fa5pro_btn211->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::REGULAR, Fa5pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa5pro_btn212->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::REGULAR, Fa5pro::fa_star,    QColor(214, 197,  64)));
        ui->fa5pro_btn213->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::REGULAR, Fa5pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa5pro_btn214->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::REGULAR, Fa5pro::fa_map,     QColor(102, 163,  52)));
        ui->fa5pro_btn215->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::REGULAR, Fa5pro::fa_comment, QColor(  0,   0,   0)));

        ui->fa5pro_btn221->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::LIGHT,   Fa5pro::fa_flag,    QColor(195,  65,  65)));
        ui->fa5pro_btn222->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::LIGHT,   Fa5pro::fa_star,    QColor(214, 197,  64)));
        ui->fa5pro_btn223->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::LIGHT,   Fa5pro::fa_user,    QColor( 69, 100, 214)));
        ui->fa5pro_btn224->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::LIGHT,   Fa5pro::fa_map,     QColor(102, 163,  52)));
        ui->fa5pro_btn225->setIcon(ZFontIcon::icon(Fa5pro::FAMILY, Fa5pro::LIGHT,   Fa5pro::fa_comment, QColor(  0,   0,   0)));
    }
    else {
        // Remove FA5 Pro tab
        QWidget *fa5pro_tab= ui->tabWidget->findChild<QWidget*>("fa5pro_tab");
        int tabIndex= ui->tabWidget->indexOf(fa5pro_tab);
        ui->tabWidget->removeTab(tabIndex);
    }


    //-----------------------------------
    // Font Awesome 4.7.0 FREE (Legacy) -
    //-----------------------------------
    ui->fa4_btn101->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_flag,      QColor(195,  65,  65)));
    ui->fa4_btn102->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_star,      QColor(214, 197,  64)));
    ui->fa4_btn103->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_user,      QColor( 69, 100, 214)));
    ui->fa4_btn104->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_map,       QColor(102, 163,  52)));
    ui->fa4_btn105->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_comment,   QColor(  0,   0,   0)));

    ui->fa4_btn111->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_flag_o,    QColor(195,  65,  65)));
    ui->fa4_btn112->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_star_o,    QColor(214, 197,  64)));
    ui->fa4_btn113->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_user_o,    QColor( 69, 100, 214)));
    ui->fa4_btn114->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_map_o,     QColor(102, 163,  52)));
    ui->fa4_btn115->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_comment_o, QColor(  0,   0,   0)));

    ui->fa4_btn121->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_github,    QColor(0, 0, 0)));
    ui->fa4_btn122->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_gitlab,    QColor(0, 0, 0)));
    ui->fa4_btn123->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_facebook,  QColor(0, 0, 0)));
    ui->fa4_btn124->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_twitter,   QColor(0, 0, 0)));
    ui->fa4_btn125->setIcon(ZFontIcon::icon(Fa4::FAMILY, Fa4::fa_instagram, QColor(0, 0, 0)));


    //---------------------------------------------
    // Google Material Design Icons 4.x | 96206ad -
    //---------------------------------------------
    ui->gmi_btn101->setIcon(ZFontIcon::icon(Gmi4::FAMILY_OUTLINED, Gmi4::gmi_tour,        QColor(195,  65,  65)));
    ui->gmi_btn102->setIcon(ZFontIcon::icon(Gmi4::FAMILY_OUTLINED, Gmi4::gmi_star_rate,   QColor(214, 197,  64)));
    ui->gmi_btn103->setIcon(ZFontIcon::icon(Gmi4::FAMILY_OUTLINED, Gmi4::gmi_account_box, QColor( 69, 100, 214)));
    ui->gmi_btn104->setIcon(ZFontIcon::icon(Gmi4::FAMILY_OUTLINED, Gmi4::gmi_map,         QColor(102, 163,  52)));
    ui->gmi_btn105->setIcon(ZFontIcon::icon(Gmi4::FAMILY_OUTLINED, Gmi4::gmi_comment,     QColor(  0,   0,   0)));

    ui->gmi_btn111->setIcon(ZFontIcon::icon(Gmi4::FAMILY_ROUND,    Gmi4::gmi_tour,        QColor(195,  65,  65)));
    ui->gmi_btn112->setIcon(ZFontIcon::icon(Gmi4::FAMILY_ROUND,    Gmi4::gmi_star_rate,   QColor(214, 197,  64)));
    ui->gmi_btn113->setIcon(ZFontIcon::icon(Gmi4::FAMILY_ROUND,    Gmi4::gmi_account_box, QColor( 69, 100, 214)));
    ui->gmi_btn114->setIcon(ZFontIcon::icon(Gmi4::FAMILY_ROUND,    Gmi4::gmi_map,         QColor(102, 163,  52)));
    ui->gmi_btn115->setIcon(ZFontIcon::icon(Gmi4::FAMILY_ROUND,    Gmi4::gmi_comment,     QColor(  0,   0,   0)));

    ui->gmi_btn121->setIcon(ZFontIcon::icon(Gmi4::FAMILY_SHARP,    Gmi4::gmi_tour,        QColor(195,  65,  65)));
    ui->gmi_btn122->setIcon(ZFontIcon::icon(Gmi4::FAMILY_SHARP,    Gmi4::gmi_star_rate,   QColor(214, 197,  64)));
    ui->gmi_btn123->setIcon(ZFontIcon::icon(Gmi4::FAMILY_SHARP,    Gmi4::gmi_account_box, QColor( 69, 100, 214)));
    ui->gmi_btn124->setIcon(ZFontIcon::icon(Gmi4::FAMILY_SHARP,    Gmi4::gmi_map,         QColor(102, 163,  52)));
    ui->gmi_btn125->setIcon(ZFontIcon::icon(Gmi4::FAMILY_SHARP,    Gmi4::gmi_comment,     QColor(  0,   0,   0)));
}

Dialog::~Dialog() {
    delete ui;
}
